### PR TITLE
e2e/storage: avoid definining unusable tests

### PR DESCRIPTION
**Pull request by pohly**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

After the recent `test/e2e/storage` refactoring (merged for 1.13), a lot of tests get defined that can only run with specific cluster providers. Those tests then get skipped at runtime. Even tests that get skipped take time (test setup, creating client set or worse, deploying the driver if the driver needs that) and the list of skipped tests in the CI grew a lot without adding any information (skip reason not captured). Therefore this PR:
- enables the definition of tests depending on the test context
- changes the "skip test" API so that test definition gets skipped, while still allowing runtime skipping
- changes the CSI and in-tree drivers to skip definition as much as possible

**Special notes for your reviewer**:

This contains PR #70862 because of code conflicts, so that PR must be merged first.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

/sig storage
/cc @mkimuram 
/assign @msau42

### CODE REVIEW

**Comment by pohly** [comment]
/hold 

PR #70862 must be merged first.

**Comment by pohly** [comment]
There's one change that I didn't make to avoid unnecessary churn: `testsuites.RunTestSuite` isn't actually *running* any tests, it merely *defines* them. This is an important distinction for people reading the code or trying to understand how to use the API, because they have to understand that defining tests and executing them happens at entirely different points in time.

"test suite" is also ambiguous, because in the larger context of the final test binary, the entire set of specs is the "test suite". What the method adds is merely a certain set of tests (or even more precisely, some "specs" in Ginkgo terms).

Therefore it would be useful to rename the method to "AddTests" or "DefineTests". I'd prefer to not use the term "specs" because it is Ginkgo-specific and doesn't match at all with the naming of the package ("testsuites").

Anyway, let me know what you think about this. I can still add the renaming to this PR. But if we do it, we should do it now, to avoid breaking external tests suites which might start using the package.

File: test/e2e/storage/drivers/in_tree.go [hunk 1]
```diff
@@ -260,11 +258,16 @@ func (g *glusterFSDriver) GetDriverInfo() *testsuites.DriverInfo {
...
+	if !framework.NodeOSDistroIs("gci", "ubuntu", "custom") {
+		return true
+	}
 	if pattern.FsType == "xfs" {
-		framework.SkipUnlessNodeOSDistroIs("ubuntu", "custom")
+		if !framework.NodeOSDistroIs("ubuntu", "custom") {
+			return true
+		}
 	}
+	return false
```
Comment (verult) [inline test/e2e/storage/drivers/in_tree.go]: nit: 
```golang
!framework.NodeOSDistroIs("gci", "ubuntu", "custom") || 
(pattern.FsType == "xfs" && 
!framework.NodeOSDistroIs("ubuntu", "custom"))
```

(negated if you are changing the function name)

File: test/e2e/storage/drivers/in_tree.go [hunk 2]
```diff
@@ -1187,11 +1174,16 @@ func (g *gcePdDriver) GetDriverInfo() *testsuites.DriverInfo {
...
+	if !framework.ProviderIs("gce", "gke") {
+		return true
+	}
 	if pattern.FsType == "xfs" {
-		framework.SkipUnlessNodeOSDistroIs("ubuntu", "custom")
+		if !framework.NodeOSDistroIs("ubuntu", "custom") {
+			return true
+		}
 	}
+	return false
```
Comment (verult) [inline test/e2e/storage/drivers/in_tree.go]: nit: also convert to boolean form

File: test/e2e/storage/testsuites/base.go [hunk 3]
```diff
@@ -64,26 +64,37 @@ func getTestNameStr(suite TestSuite, pattern testpatterns.TestPattern) string {
...
-// is not suitable to be tested.
+// skipUnsupportedTest will determine if the combination of driver, testsuite, and testpattern
+// is unsuitable to be tested.
 // Whether it needs to be skipped is checked by following steps:
 // 1. Check if Whether volType is supported by driver from its interface
 // 2. Check if fsType is supported by driver
 // 3. Check with driver specific logic
 // 4. Check with testSuite specific logic
-func skipUnsupportedTest(suite TestSuite, driver TestDriver, pattern testpatterns.TestPattern) {
+func skipUnsupportedTest(suite TestSuite, driver TestDriver, pattern testpatterns.TestPattern) bool {
```
Comment (verult) [inline test/e2e/storage/testsuites/base.go]: Consider renaming this too

File: test/e2e/storage/testsuites/base.go [hunk 4]
```diff
@@ -64,26 +64,37 @@ func getTestNameStr(suite TestSuite, pattern testpatterns.TestPattern) string {
...
-// RunTestSuite runs all testpatterns of all testSuites for a driver
+// RunTestSuite defines tests for all testpatterns and all testSuites for a driver
 func RunTestSuite(f *framework.Framework, driver TestDriver, tsInits []func() TestSuite, tunePatternFunc func([]testpatterns.TestPattern) []testpatterns.TestPattern) {
 	for _, testSuiteInit := range tsInits {
 		suite := testSuiteInit()
 		patterns := tunePatternFunc(suite.getTestSuiteInfo().testPatterns)
 
 		for _, pattern := range patterns {
+			if skipUnsupportedTest(suite, driver, pattern) {
+				continue
```
Comment (verult) [inline test/e2e/storage/testsuites/base.go]: Log then skip. The user should know a particular test is detected but skipped due to driver incompatibility.

File: test/e2e/storage/testsuites/volumes.go [hunk 5]
```diff
@@ -72,21 +72,18 @@ func (t *volumesTestSuite) getTestSuiteInfo() TestSuiteInfo {
 	return t.tsInfo
 }
 
-func (t *volumesTestSuite) skipUnsupportedTest(pattern testpatterns.TestPattern, driver TestDriver) {
+func (t *volumesTestSuite) skipUnsupportedTest(pattern testpatterns.TestPattern, driver TestDriver) bool {
+	return false
 }
 
-func skipPersistenceTest(driver TestDriver) {
+func skipPersistenceTest(driver TestDriver) bool {
```
Comment (verult) [inline test/e2e/storage/testsuites/volumes.go]: also seems like a double negative

File: test/e2e/storage/testsuites/volumes.go [hunk 6]
```diff
@@ -72,21 +72,18 @@ func (t *volumesTestSuite) getTestSuiteInfo() TestSuiteInfo {
...
 	dInfo := driver.GetDriverInfo()
-	if !dInfo.Capabilities[CapPersistence] {
-		framework.Skipf("Driver %q does not provide persistency - skipping", dInfo.Name)
-	}
+	return !dInfo.Capabilities[CapPersistence]
 }
 
-func skipExecTest(driver TestDriver) {
+func skipExecTest(driver TestDriver) bool {
```
Comment (verult) [inline test/e2e/storage/testsuites/volumes.go]: ditto

File: test/e2e/storage/utils/framework.go [hunk 7]
```diff
@@ -16,8 +16,12 @@ limitations under the License.
...
-import "github.com/onsi/ginkgo"
+import "k8s.io/kubernetes/test/e2e/framework"
 
+// SIGDescribe adds the sig-storage tag and ensures that the body only
+// gets called after TestContext is valid. Therefore it is okay to use
+// functions like framework.NodeDistroOSIs or framework.ProviderIs to
+// add specs conditionally.
 func SIGDescribe(text string, body func()) bool {
-	return ginkgo.Describe("[sig-storage] "+text, body)
+	return framework.DescribeWithTestContext("[sig-storage] "+text, body)
```
Comment (verult) [inline test/e2e/storage/utils/framework.go]: /cc @rmmh

File: test/e2e/framework/specs.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+var specs []spec
+
+// DescribeWithTestContext is like ginkgo.Describe, but in contrast to
+// ginkgo, the framework implementation will invoke the spec body at a
+// time when TestContext has been fully initialized. A callback
+// invoked via framework.Describe therefore can add or parameterize
+// specs depending on the test configuration.
+func DescribeWithTestContext(text string, body func()) bool {
```
Comment (verult) [inline test/e2e/framework/specs.go]: /cc @rmmh

**Review by verult — COMMENTED** [review]
Reviewed the first 2 of the commits. Thanks for improving the skip process!

File: test/e2e/storage/testsuites/testdriver.go [hunk 9]
```diff
@@ -29,12 +29,32 @@ type TestDriver interface {
...
 	GetDriverInfo() *DriverInfo
 	// CreateDriver creates all driver resources that is required for TestDriver method
-	// except CreateVolume
+	// except CreateVolume. May be called more than once and should only do something on
+	// the first call.
 	CreateDriver()
-	// CreateDriver cleanup all the resources that is created in CreateDriver
+	// CreateDriver cleanup all the resources that is created in CreateDriver. There is
+	// no guarantee that CreateDriver succeeded or even was called at all, so the test driver
+	// has to track resources.
```
[the remark this answers is not in the document]
Comment (pohly) [reply test/e2e/storage/testsuites/testdriver.go]: CleanupDriver() will always get called, under all circumstances. This has been true also before, it just wasn't explicitly mentioned, therefore I added this comment.

File: test/e2e/storage/testsuites/testdriver.go [hunk 10]
```diff
@@ -29,12 +29,32 @@ type TestDriver interface {
...
 	CleanupDriver()
-	// SkipUnsupportedTest skips test in Testpattern is not suitable to test with the TestDriver
-	SkipUnsupportedTest(testpatterns.TestPattern)
+}
+
+// StaticSkipTestDriver is an optional interface that drivers can
+// implement to filter out unsuitable tests.
```
[the remark this answers is not in the document]
Comment (pohly) [reply test/e2e/storage/testsuites/testdriver.go]: I've renamed the interface to `FilterTestDriver` (seemed to make more sense, now that the method is called `IsTestSupported`) and added the comment.

File: test/e2e/storage/drivers/in_tree.go (hunk 1 above)
Comment (pohly) [reply test/e2e/storage/drivers/in_tree.go]: Yep.

File: test/e2e/storage/testsuites/base.go (hunk 4 above)
Comment (pohly) [reply test/e2e/storage/testsuites/base.go]: That would make each test suite invocation very verbose, because that information would get printed before running any test, whether the user intends to run the filtered out test or not.

I agree that this is useful information, but we should make it depend on some non-default command line flag. Perhaps we can (ab)use `-ginkgo.dryRun` for this?

File: test/e2e/storage/testsuites/volumes.go (hunk 5 above)
Comment (pohly) [reply test/e2e/storage/testsuites/volumes.go]: I've inlined the code from these functions and removed them. They were trivial and now `testVolumes` is easier to understand.

**Comment by pohly** [comment]
@verult pushed, please have another look.

File: test/e2e/storage/testsuites/testdriver.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (verult) [reply test/e2e/storage/testsuites/testdriver.go]: To me, the new comment seems to imply the behavior of CleanupDriver() is different depending on what happens with CreateDriver(). Could you improve the comment to make what you said here more explicit? (Maybe ".... *and* the test driver has to *clean up resources regardless*")

File: test/e2e/storage/testsuites/base.go (hunk 4 above)
Comment (verult) [reply test/e2e/storage/testsuites/base.go]: To further reduce the verbosity, we could print a single log message per test suite and accumulate incompatible drivers in a list. This is still a huge improvement over the verbosity we have today.

I'm not a fan of relying on a non-default flag that could potentially be hard to discover, or in the case of using `-ginkgo.dryRun`, hard to correlate with controlling the log level here.

File: test/e2e/storage/testsuites/testdriver.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (pohly) [reply test/e2e/storage/testsuites/testdriver.go]: Okay.

File: test/e2e/storage/drivers/in_tree.go [hunk 11]
```diff
@@ -260,11 +258,16 @@ func (g *glusterFSDriver) GetDriverInfo() *testsuites.DriverInfo {
...
+	if !framework.NodeOSDistroIs("gci", "ubuntu", "custom") {
+		return false
+	}
 	if pattern.FsType == "xfs" {
-		framework.SkipUnlessNodeOSDistroIs("ubuntu", "custom")
+		if !framework.NodeOSDistroIs("ubuntu", "custom") {
+			return false
+		}
 	}
+	return true
```
Comment (verult) [inline test/e2e/storage/drivers/in_tree.go]: nit:

Something like this is arguably easier to read:

```golang
if framework.NodeOSDistroIs("gci", "ubuntu", "custom") {
    if pattern.FsType != "xfs" || framework.NodeOSDistroIs("ubuntu", "custom") {
        return true
    }
}
return false
```
To improve it further, I would personally use this (but up to you):
```golang
return framework.NodeOSDistroIs("gci", "ubuntu", "custom") &&
       (pattern.FsType != "xfs" ||
        framework.NodeOSDistroIs("ubuntu", "custom"))
```

The double negatives in your code made it a bit hard for me to parse

File: test/e2e/storage/drivers/in_tree.go [hunk 12]
```diff
@@ -1187,11 +1174,16 @@ func (g *gcePdDriver) GetDriverInfo() *testsuites.DriverInfo {
...
+	if !framework.ProviderIs("gce", "gke") {
+		return false
+	}
 	if pattern.FsType == "xfs" {
-		framework.SkipUnlessNodeOSDistroIs("ubuntu", "custom")
+		if !framework.NodeOSDistroIs("ubuntu", "custom") {
+			return false
+		}
 	}
+	return true
```
Comment (verult) [inline test/e2e/storage/drivers/in_tree.go]: ditto

File: test/e2e/storage/testsuites/base.go (hunk 4 above)
Comment (pohly) [reply test/e2e/storage/testsuites/base.go]: I'm not sure I understand what the output should look like. How many lines should that log message have? What information do you expect?

Ideally, it would be something like:
```
The following tests were not added because they are unsupported:
[sig-storage] In-tree Volumes [Driver: nfs] [Testpattern: Dynamic PV (default fs) block mode] - block mode not supported
...
[sig-storage] CSI Volumes [Driver: gcepd] [Testpattern: Dynamic PV (default fs)] - not running on GCE
...
```

But that will be many lines of output, basically one line for each suite/pattern combination unless we implement some intelligence in RunTestSuite that determines whether all patterns or even all combinations were removed. The other problem is that the reason for skipping is not transported back to RunTestSuite. We either just print the test name prefix and no explanation, or we have to change the return type.

File: test/e2e/storage/testsuites/base.go [hunk 13]
```diff
@@ -37,8 +37,8 @@ import (
 type TestSuite interface {
 	// getTestSuiteInfo returns the TestSuiteInfo for this TestSuite
 	getTestSuiteInfo() TestSuiteInfo
-	// skipUnsupportedTest skips the test if this TestSuite is not suitable to be tested with the combination of TestPattern and TestDriver
-	skipUnsupportedTest(testpatterns.TestPattern, TestDriver)
+	// isTestSupported returns false if this TestSuite cannot be tested with the combination of TestPattern and TestDriver
```
Comment (verult) [inline test/e2e/storage/testsuites/base.go]: nit: "returns *true* if this TestSuite *can* be..."

File: test/e2e/storage/testsuites/base.go [hunk 14]
```diff
@@ -64,26 +64,38 @@ func getTestNameStr(suite TestSuite, pattern testpatterns.TestPattern) string {
...
 		}
 	}
 }
 
-// skipUnsupportedTest will skip tests if the combination of driver, testsuite, and testpattern
-// is not suitable to be tested.
+// isTestSupported will determine if the combination of driver, testsuite, and testpattern
+// can be tested.
+//
 // Whether it needs to be skipped is checked by following steps:
```
Comment (verult) [inline test/e2e/storage/testsuites/base.go]: "Whether *the test combination is supported* is checked..."

File: test/e2e/storage/testsuites/testdriver.go [hunk 15]
```diff
@@ -29,12 +29,32 @@ type TestDriver interface {
...
+	// has to track resources.
 	CleanupDriver()
-	// SkipUnsupportedTest skips test in Testpattern is not suitable to test with the TestDriver
-	SkipUnsupportedTest(testpatterns.TestPattern)
+}
+
+// FilterTestDriver is an optional interface that drivers can
+// implement to filter out unsuitable tests while tests get defined.
+type FilterTestDriver interface {
+	// IsTestSupported returns false if the Testpattern is not
```
Comment (verult) [inline test/e2e/storage/testsuites/testdriver.go]: Testpattern -> TestPattern

nit: "...returns *true* if the TestPattern is *suitable*..."

**Review by verult — COMMENTED** [review]
Posting comments for first 2 commits for now; I'm working on the last 2

File: test/e2e/storage/drivers/in_tree.go (hunk 11 above)
Comment (pohly) [reply test/e2e/storage/drivers/in_tree.go]: I also like the latter better, but again didn't want to change tooo much to keep the review simpler: when a `ginkgo.Skip` gets replaced by a `return true` or now `return true`, then it is obvious that the overall logic is unchanged.

I'll change it.

File: test/e2e/storage/testsuites/provisioning.go [hunk 16]
```diff
@@ -88,10 +88,9 @@ func createProvisioningTestInput(driver TestDriver, pattern testpatterns.TestPat
 			ClaimSize:    resource.claimSize,
 			ExpectedSize: resource.claimSize,
 		},
-		cs:    driver.GetDriverInfo().Config.Framework.ClientSet,
-		pvc:   resource.pvc,
-		sc:    resource.sc,
-		dInfo: driver.GetDriverInfo(),
```
Comment (verult) [inline test/e2e/storage/testsuites/provisioning.go]: Can we remove the field entirely?

File: test/e2e/storage/testsuites/provisioning.go [hunk 17]
```diff
@@ -120,13 +119,11 @@ func (p *provisioningTestSuite) execTest(driver TestDriver, pattern testpatterns
 		// Ginkgo's "Global Shared Behaviors" require arguments for a shared function
 		// to be a single struct and to be passed as a pointer.
 		// Please see https://onsi.github.io/ginkgo/#global-shared-behaviors for details.
-		testProvisioning(&input)
+		testProvisioning(driver, &input)
```
Comment (verult) [inline test/e2e/storage/testsuites/provisioning.go]: Just want to make sure I understand: `BeforeEach()` runs before `It()`s, but not necessarily before `testProvisioning()`?

**Review by verult — COMMENTED** [review]
2nd half look good. Added some minor comments

File: test/e2e/storage/testsuites/provisioning.go (hunk 16 above)
Comment (pohly) [reply test/e2e/storage/testsuites/provisioning.go]: Which field? I'm sorry, but it's not obvious in the Github UI.

Comment (verult) [reply test/e2e/storage/testsuites/provisioning.go]: Np. `dInfo`

Comment (pohly) [reply test/e2e/storage/testsuites/provisioning.go]: But dInfo is gone already, I'm removing it in this PR already because it was clearly redundant.

Is there anything else you want to get removed?

File: test/e2e/storage/testsuites/base.go (hunk 4 above)
Comment (pohly) [reply test/e2e/storage/testsuites/base.go]: I tried to implement something like this, but hit a major snag: there's no way to retrieve the current set of components (i.e. the text from previous `Describe` and `Context` calls) from ginkgo during test registration. The effect is that at the place where an `It` call gets skipped, not enough information is available to record the entire test name. In my current prototype, the inner test name is available and (with some extra effort) the testsuite/pattern part. The driver name could be added, but that's still no guarantee that the resulting string is complete.

Here's what it looks like:
```
$ go test -v ./test/e2e 
W1223 17:21:55.928475   18945 test_context.go:392] Unable to find in-cluster config, using default host : http://127.0.0.1:8080
Dec 23 17:21:55.928: INFO: The --provider flag is not set.  Treating as a conformance test.  Some tests may not be run.
I1223 17:21:56.041325   18945 test_context.go:427] The following tests were not added to the test suite because they are not supported in the current configuration:
??? [Testpattern: Dynamic PV (default fs)] provisioning should provision storage with mount options
??? [Testpattern: Dynamic PV (default fs)] provisioning should provision storage with mount options
??? [Testpattern: Dynamic PV (default fs)] provisioning should provision storage with mount options
??? [Testpattern: Dynamic PV (default fs)] provisioning should provision storage with mount options
??? should create and delete block persistent volumes
??? should create and delete block persistent volumes
??? should create and delete block persistent volumes
??? should create and delete block persistent volumes
??? should create and delete block persistent volumes
```

The prototype commit is here: https://github.com/pohly/kubernetes/commit/20e270a695b13d89a7d13337ab96aa9f0d65eacb

I see two options:
1. take this PR as it is, without logging tests skipped during test registration
2. work with Ginkgo upstream such that it better supports skipped test registration

The latter will take time. Ginkgo is not entirely frozen, but maintenance isn't particularly active either. One of my PRs has been waiting to get merged for a while now.

@verult @msau42: I'll let you guys decide.

File: test/e2e/storage/testsuites/testdriver.go (hunk 15 above)
Comment (pohly) [reply test/e2e/storage/testsuites/testdriver.go]: Done.

File: test/e2e/storage/testsuites/base.go (hunk 14 above)
Comment (pohly) [reply test/e2e/storage/testsuites/base.go]: Done.

File: test/e2e/storage/testsuites/base.go (hunk 13 above)
Comment (pohly) [reply test/e2e/storage/testsuites/base.go]: Done.

File: test/e2e/storage/drivers/in_tree.go (hunk 11 above)
Comment (pohly) [reply test/e2e/storage/drivers/in_tree.go]: Done.

File: test/e2e/storage/drivers/in_tree.go (hunk 12 above)
Comment (pohly) [reply test/e2e/storage/drivers/in_tree.go]: Done.

File: test/e2e/storage/testsuites/provisioning.go (hunk 17 above)
Comment (pohly) [reply test/e2e/storage/testsuites/provisioning.go]: PR https://github.com/kubernetes/kubernetes/pull/72434 addresses this.

**Comment by pohly** [comment]
I think I addressed all review feedback except the open question around logging of tests that don't get registered (https://github.com/kubernetes/kubernetes/pull/70992#discussion_r243769787) and rebased onto current master.

I put the logging topic onto the agenda for the next SIG-storage meeting on January 8th.

File: test/e2e/storage/testsuites/provisioning.go (hunk 16 above)
Comment (pohly) [reply test/e2e/storage/testsuites/provisioning.go]: https://github.com/kubernetes/kubernetes/pull/72434 removes the entire struct holding these fields.

File: test/e2e/storage/testsuites/testdriver.go [hunk 18]
```diff
@@ -29,12 +29,32 @@ type TestDriver interface {
...
+type FilterTestDriver interface {
+	// IsTestSupported returns true if the Testpattern is
+	// suitable to test with the TestDriver. This will be called
+	// already while defining tests and unsupported tests will not even
+	// be added to the test suite.
+	IsTestSupported(testpatterns.TestPattern) bool
+}
+
+// BeforeEachTestDriver is an optional interface that drivers can
+// implement to hook into test execution. It can be used to initialize
```
Comment (mkimuram) [inline test/e2e/storage/testsuites/testdriver.go]: Do we need to allow this method to initialize additional resources?
Are there any existing use cases for it?

(I'm worrying about that it was due to my previous mistake in putting resource initialization code for GCE driver here, which has already been moved to right place in another commit.)

**Review by mkimuram — COMMENTED** [review]
@pohly

Sorry for late response.

I still don't completely understand the logic behind ```DescribeWithTestContext```, the other logics look good to me if it works correctly.
Other than above, let me add minor comments for naming and comment.

File: test/e2e/storage/testsuites/testdriver.go [hunk 19]
```diff
@@ -29,12 +29,32 @@ type TestDriver interface {
...
-	// CreateDriver cleanup all the resources that is created in CreateDriver
+	// CreateDriver cleanup all the resources that is created in CreateDriver. There is
+	// no guarantee that CreateDriver succeeded or even was called at all, so the test driver
+	// has to track resources.
 	CleanupDriver()
-	// SkipUnsupportedTest skips test in Testpattern is not suitable to test with the TestDriver
-	SkipUnsupportedTest(testpatterns.TestPattern)
+}
+
+// FilterTestDriver is an optional interface that drivers can
```
[the remark this answers is not in the document]
Comment (pohly) [reply test/e2e/storage/testsuites/testdriver.go]: Do you have a better name for FilterTestDriver? I'm completely open for suggestions here.

File: test/e2e/storage/testsuites/testdriver.go (hunk 18 above)
Comment (pohly) [reply test/e2e/storage/testsuites/testdriver.go]: It's not really needed and in fact later (https://github.com/kubernetes/kubernetes/pull/72002) I ended up removing it again because it didn't turn out to be that useful. CreateDriver gets called at the same time and can replace the extra `BeforeEachTestDriver`.

I'll remove it directly in this PR instead.

Comment (pohly) [reply test/e2e/storage/testsuites/testdriver.go]: Done.

File: test/e2e/storage/testsuites/testdriver.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (mkimuram) [reply test/e2e/storage/testsuites/testdriver.go]: @pohly 

Now that ```BeforeEachTestDriver``` is deleted, so there is no need to compare and distinguish these two for whether they are runtime or not. In this situation, I think that current naming is good enough.

Thank you for your consideration and handling on this.

File: test/e2e/framework/specs.go [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+)
+
+type spec struct {
+	text string
+	body func()
+}
+
+var specs []spec
+
+// DescribeWithTestContext is like ginkgo.Describe, but in contrast to
```
Comment (wongma7) [inline test/e2e/framework/specs.go]: I would like the comment to mention the motivation for why someone might want to use this function to "invoke spec bod[ies]" only after "TestContext has been fully initialized". i.e. this exists so that we can skip tests in spec bodies based on TestContext. Otherwise the file is too generic looking and the technical language imposing :]

Comment (wongma7) [reply test/e2e/framework/specs.go]: I understand "can add or parameterize specs depending on the test configuration" basically explains it but I would word it differently is all.

**Comment by pohly** [comment]
In the SIG-storage meeting today, the first question was why skipping tests at runtime is considered bad. I pointed out that drivers get deployed for each test, for which the reply was that a) the driver installation should be triggered when a test really runs (which I don't think can be done nicely) or b) driver installation should happen before the test suite runs.

However, it's not clear how to get the driver installation hooked into the cluster setup. This will require further discussion.

But even when driver installation is not part of the test setup, skipped tests still take time. I don't have hard numbers about that, though. Let's take the gcepd production driver as example, which is pre-installed. @msau42 what's your opinion, do you prefer skipping unusable tests for that driver at runtime or at testsuite definition time?

File: test/e2e/framework/specs.go (hunk 20 above)
Comment (pohly) [reply test/e2e/framework/specs.go]: I had another use case in mind: testing CSI drivers with the upstream e2e test binary.

With this change, the e2e test binary could have a command line parameter(s) that enable testing of a driver by providing the necessary information. The "TestDriver" instance that is given to the storage test suite then gets defined based on these parameters.

Without this change, the same can still be done, but would have to be based on parsing environment variables.

**Comment by msau42** [comment]
> But even when driver installation is not part of the test setup, skipped tests still take time. I don't have hard numbers about that, though.

Could you try running locally some non-supported environment tests, with only the change that adds skip to `CreateDriver`, to measure how much just skipping the tests take?

**Comment by pohly** [comment]
That can already be tested without any changes. On a 1.13 cluster
brought up with `hack/local-up-cluster.sh`, running all tests involving
the external gcepd driver takes:

```
$ go test -v ./test/e2e -args -ginkgo.v -ginkgo.focus=Feature..gcePD-external
...
Will run 31 of 2161 specs
...

S [SKIPPING] in Spec Setup (BeforeEach) [6.157 seconds]
[sig-storage] CSI Volumes
/nvme/gopath/src/k8s.io/kubernetes/test/e2e/storage/utils/framework.go:22
  [Driver: pd.csi.storage.gke.io][Feature: gcePD-external]
  /nvme/gopath/src/k8s.io/kubernetes/test/e2e/storage/csi_volumes.go:129
    [Testpattern: Dynamic PV (ext3)] volumes
    /nvme/gopath/src/k8s.io/kubernetes/test/e2e/storage/testsuites/volumes.go:128
      should be mountable [BeforeEach]
      /nvme/gopath/src/k8s.io/kubernetes/test/e2e/storage/testsuites/volumes.go:169

      Jan  9 06:01:46.774: Only supported for providers [gce gke] (not )

      /nvme/gopath/src/k8s.io/kubernetes/test/e2e/framework/util.go:292
...
Ran 0 of 2161 Specs in 190.920 seconds
SUCCESS! -- 0 Passed | 0 Failed | 0 Pending | 2161 Skipped --- PASS: TestE2E (191.01s)
```

That's 6 seconds per test. Not too bad and 3 minutes overall don't matter in a CI setup. It could be more annoying when running the tests interactively, though.

But as soon as `CreateDriver` is expensive, it gets worse of course. We might be able to solve this (eventually) by moving the driver installation into the test cluster setup phase, but what about drivers like NFS which bring up a NFS server for each test? Should and can that also be part of the cluster setup?

It might be easier to delay the `CreateDriver` call. Right now it gets called outside of the test suites. If we make the test suites responsible for calling it right before running a specific test, then runtime skipping of tests that aren't supported by the driver + test pattern combination becomes cheap.

**Comment by pohly** [comment]
Even better, if the framework instantiation also gets moved after checking the driver + test pattern combination, then we can also save those 6 seconds.

**Comment by pohly** [comment]
There is value in skipping tests at runtime (consistency, result reporting via Ginkgo). By reordering the initialization steps it is possible to address the main pain point that this PR was meant for (avoiding useless driver deployment in skipped tests), see PR #72434. Therefore I am closing this PR.
